base/acme/src/org/dogtagpki/acme/backend/PKIBackend.java

```diff
@@ -0,0 +1,283 @@
+import com.netscape.certsrv.ca.CACertClient;
+import com.netscape.certsrv.ca.CAClient;
+import com.netscape.certsrv.cert.CertData;
+import com.netscape.certsrv.cert.CertEnrollmentRequest;
+import com.netscape.certsrv.cert.CertRequestInfo;
+import com.netscape.certsrv.cert.CertRequestInfos;
+import com.netscape.certsrv.cert.CertReviewResponse;
+import com.netscape.certsrv.cert.CertRevokeRequest;
+import com.netscape.certsrv.client.ClientConfig;
+import com.netscape.certsrv.client.PKIClient;
+import com.netscape.certsrv.dbs.certdb.CertId;
+import com.netscape.certsrv.profile.ProfileAttribute;
+import com.netscape.certsrv.profile.ProfileInput;
+import com.netscape.certsrv.request.RequestId;
+import com.netscape.certsrv.request.RequestStatus;
+import com.netscape.cmsutil.crypto.CryptoUtil;
+
+/**
+ * @author Endi S. Dewata
+ */
+public class PKIBackend extends ACMEBackend {
+
+    public static org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(PKIBackend.class);
+
+    private String url;
+    private String profile;
+    private String nickname;
+    private String username;
+    private String password;
+
+    public String getURL() {
+        return url;
+    }
+
+    public void setURL(String url) {
+        this.url = url;
+    }
+
+    public String getProfile() {
+        return profile;
+    }
+
+    public void setProfile(String profile) {
+        this.profile = profile;
+    }
+
+    public String getNickname() {
+        return nickname;
+    }
+
+    public void setNickname(String nickname) {
+        this.nickname = nickname;
+    }
+
+    public String getUsername() {
+        return username;
+    }
+
+    public void setUsername(String username) {
+        this.username = username;
+    }
+
+    public String getPassword() {
+        return password;
+    }
+
+    public void setPassword(String password) {
+        this.password = password;
+    }
+
+    public void init() {
+        url = config.getParameter("url");
+        profile = config.getParameter("profile");
+        nickname = config.getParameter("nickname");
+        username = config.getParameter("username");
+        password = config.getParameter("password");
+    }
+
+    public String issueCertificate(String csr) throws Exception {
+
+        logger.info("Issuing certificate");
+
+        logger.info("PKI server: " + url);
+        logger.info("- profile: " + profile);
+        if (nickname != null) {
+            logger.info("- nickname: " + nickname);
+        }
+        if (username != null) {
+            logger.info("- username: " + username);
+        }
+
+        AuthorityID aid = null;
+        X500Name adn = null;
+
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.setServerURL(url);
+        clientConfig.setCertNickname(nickname);
+        clientConfig.setUsername(username);
+        clientConfig.setPassword(password);
+
+        PKIClient pkiClient = new PKIClient(clientConfig);
```

Comment:
Are there any reasons not to make `PKIClient` a member of `PKIBackend`? You use one in all three methods. It seems to me that your ideal behavior would be to load the configuration as early as possible on server startup and refuse to start if the config is bad.

I guess there might be timeouts, but I wouldn't imagine you'd need to lock the `CACertClient` once it is set up, would you? At any rate, could you push `PKIClient` and `CAClient` into the object and make only `CACertClient` instance variables when you need to issue a specific operations?

Reply:
Yeah, I think we can do that. Let me update the patch.